### PR TITLE
[BACKLOG-7998] - Improve perfomance of 'Total' functions

### PR DESCRIPTION
**Pull request by EgorZhuk**
Finishing @prakapenka work.
Junits are on the way - need to see wingman's opinion.
@tmorgner

### CODE REVIEW

**Comment by EgorZhuk** [comment]
Oh, need to make StateSequence Serializable

File: engine/core/source/org/pentaho/reporting/engine/classic/core/function/StateSequence.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+package org.pentaho.reporting.engine.classic.core.function;
+
+import org.pentaho.reporting.engine.classic.core.states.ReportStateKey;
+import org.pentaho.reporting.engine.classic.core.util.BulkArrayList;
+import org.pentaho.reporting.engine.classic.core.util.Sequence;
+
+import java.io.Serializable;
+
+class StateSequence<T> implements Serializable {
```
Comment (tmorgner) [inline engine/core/source/org/pentaho/reporting/engine/classic/core/function/StateSequence.java]: Please make this class and its constructors public - it is a most helpful addition for anyone who wants to implement their own total-*-functions anywhere. OEMs love that stuff :)

**Comment by EgorZhuk** [comment]
@tmorgner I've added a couple of junits for classfactory. Checked what is not covered in TotalGroupSumFunction - https://yadi.sk/i/DnOej471s8sJt Is it possible that global state key equals group key?

**Comment by EgorZhuk** [comment]
Newline added

**Comment by tmorgner** [comment]
> @tmorgner I've added a couple of junits for classfactory. Checked what is not covered in TotalGroupSumFunction - https://yadi.sk/i/DnOej471s8sJt Is it possible that global state key equals group key?

No. StateKey equality depends on the "stateCode" value. For global keys, it will be ReportEvent.REPORT_INITIALIZED, for group starts it will always be GROUP_STARTED. Under no (valid) circumstances can you run into that line.

**Comment by tmorgner** [comment]
An easy way to test the TotalGroupSumFunction would probably be a mocked-up test. For the event, you will need a ReportState with a report and some data. You can use a dummy master-report with just some suitable groups to cover it. 

> RelationalGroup g = new RelationalGroup();
> g.setName("myGroup");
> g.addField("field-A");
>  MasterReport.addGroup(g);

The function only uses simple getters on the state, so you can get away with a very simple mock for it. DataRow has easy to use standalone implementations (StaticDataRow), so there is no pain here either.

**Comment by tmorgner** [comment]
Good enough. Dont care about Wingmans test coverage randomness here.

**Comment by tmorgner** [comment]
@EgorZhuk If your StateSequenceList is serializable with transient members, the list will have null for these members after deserialization. Is that really what you want here?

You probably need a readResolve(..) method here to restore the BulkArrayLists to some empty objects.

(serialization wont happen while the report is running, so restoring to an empty state here is fine.)

**Comment by EgorZhuk** [comment]
@tmorgner nice catch, thank you, added it

**Pull request merged**